File: ADCToolbox_Python/test/test_batch_data.py
"""
Test batch data processing - Python version of test_batch_data.m

Tests specPlot and specPlotPhase with batch/multi-run data.
"""

import os
import sys
import numpy as np
import matplotlib.pyplot as plt

# Add project root to sys.path if needed (for direct script execution)
_project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from ADCToolbox_Python.spec_plot import spec_plot
from ADCToolbox_Python.specPlotPhase import spec_plot_phase


def main():
    """Test batch data processing."""

    # Configuration
    input_dir = os.path.join(_project_root, "ADCToolbox_example_data")
    output_dir = os.path.join(_project_root, "ADCToolbox_example_output")

    # Define files to process
    files_list = [
        'batch_sinewave_Nrun_2.csv',
        'batch_sinewave_Nrun_16.csv',
        'batch_sinewave_Nrun_100.csv'
    ]

    # Check if output directory exists
    os.makedirs(output_dir, exist_ok=True)

    # Processing loop
    print(f"Starting batch processing ({len(files_list)} files)...")
    for current_filename in files_list:
        print(f"Processing file: {current_filename}")

        # Load data
        data_file_path = os.path.join(input_dir, current_filename)
        read_data = np.loadtxt(data_file_path, delimiter=',')

        # Extract file name for subfolder
        name = os.path.splitext(current_filename)[0]
        title_string = name.replace('_', '\_')  # Escape underscore for plot title

        # Create subfolder
        sub_folder = os.path.join(output_dir, name)
        os.makedirs(sub_folder, exist_ok=True)

        # ----------------------------------------------------------------
        # PLOT 1: SPECTRUM
        # ----------------------------------------------------------------
        plt.figure(figsize=(12, 8))

        # Standard Spectrum Plot
        spec_plot(read_data, label=1, harmonic=0, OSR=1, coAvg=0)
        plt.title(f'Spectrum Plot: {title_string}', fontsize=12)

        # Save Plot 1
        output_filename1 = f'Spectrum_of_{name}_python.png'
        output_filepath1 = os.path.join(sub_folder, output_filename1)
        plt.savefig(output_filepath1, dpi=150)
        print(f'[Saved image 1] -> [{output_filepath1}]')
        plt.close()

        # ----------------------------------------------------------------
        # PLOT 2: PHASE
        # ----------------------------------------------------------------
        plt.figure(figsize=(8, 8))

        # Phase Plot
        spec_plot_phase(read_data, harmonic=50, show_plot=False)
        plt.title(f'Phase Plot: {title_string}', fontsize=12)

        # Save Plot 2
        output_filename2 = f'Phase_of_{name}_python.png'
        output_filepath2 = os.path.join(sub_folder, output_filename2)
        plt.savefig(output_filepath2, dpi=150)
        print(f'[Saved image 2] -> [{output_filepath2}]\n')
        plt.close()

    print("Batch processing complete.")


if __name__ == "__main__":
    main()

File: matlab/test/unit/test_batch_data.m
%% ADC Diagnostic Batch Processing Script (Optimized File Structure)
% Goal: Read CSV data, process it, generate two separate plots, 
%       and save both into a DEDICATED SUBFOLDER per input file.

close all; clc; clear; 

% --- 1. Configuration ---
inputDir = "ADCToolbox_example_data";
outputDir = "ADCToolbox_example_output";

% Define the list of files to process
filesList = {
    'batch_sinewave_Nrun_2.csv';
    'batch_sinewave_Nrun_16.csv'; 
    'batch_sinewave_Nrun_100.csv'; 
};

% Check if the main output directory exists, create it if not
if ~isfolder(outputDir)
    mkdir(outputDir);
end

% --- 2. Processing Loop ---
fprintf('Starting batch processing (%d files)...\n', length(filesList));
for k = 1:length(filesList)
    currentFilename = filesList{k};
    fprintf('Processing file: %s\n', currentFilename);
    
    % --- Data Loading & Processing ---
    dataFilePath = fullfile(inputDir, currentFilename);
    read_data = readmatrix(dataFilePath); 
    
    % Extract file name components for titling and creating subfolder
    [~, name, ~] = fileparts(currentFilename);
    titleString = replace(name, '_', '\_'); % Escape underscore for plot title
    
    % --- CREATE DEDICATED SUBFOLDER ---
    subFolder = fullfile(outputDir, name); % Subfolder name is the file name (without extension)
    if ~isfolder(subFolder)
        mkdir(subFolder);
    end
    
    
    % ---------------------------------------------------------------------
    % --- PLOT 1: SPECTRUM ---
    % ---------------------------------------------------------------------
    figure; % Open the first figure
    
    % Standard Spectrum Plot
    specPlot(read_data, 'label', 1, 'harmonic', 0, 'OSR', 1, 'coAvg', 0);
    title(['Spectrum Plot: ', titleString]);
    
    % Saving Plot 1 to subfolder
    % Using 'Spectrum_of_' prefix as you suggested for clarity
    outputFileName1 = ['Spectrum_of_', name, '_matlab.png'];
    outputFilePath1 = fullfile(subFolder, outputFileName1); 
    saveas(gcf, outputFilePath1); 
    fprintf('[Saved image 1] -> [%s]\n', outputFilePath1);
    close(gcf); % Close the figure 1
    
    
    % ---------------------------------------------------------------------
    % --- PLOT 2: PHASE ---
    % ---------------------------------------------------------------------
    figure; % Open the second, separate figure
    
    % Phase Plot
    specPlotPhase(read_data, 'harmonic', 50);
    title(['Phase Plot: ', titleString]);
    
    % Saving Plot 2 to subfolder
    % Using 'Phase_of_' prefix for clarity
    outputFileName2 = ['Phase_of_', name, '_matlab.png'];
    outputFilePath2 = fullfile(subFolder, outputFileName2); 
    saveas(gcf, outputFilePath2); 
    fprintf('[Saved image 2] -> [%s]\n\n', outputFilePath2);
    close(gcf); % Close the figure 2
end
fprintf('Batch processing complete.\n');